packages/peregrine/src/List/items.js

```diff
@@ -1,104 +1,137 @@
-                    isSelected={selection.has(key)}
-                    onBlur={this.handleBlur}
-                    onClick={this.getClickHandler(key)}
-                    onFocus={this.getFocusHandler(index)}
-                />
-            );
-        });
-
-        return <Fragment>{children}</Fragment>;
-    }
-
-    syncSelection() {
-        const { selection } = this.state;
-        const { onSelectionChange } = this.props;
-
-        if (onSelectionChange) {
-            onSelectionChange(selection);
+    return selection;
+};
+
+const reducer = (state, action) => {
+    switch (action.type) {
+        case 'REMOVE_FOCUS':
+            return {
+                ...state,
+                hasFocus: false
+            };
+        case 'SET_FOCUS':
+            return {
+                ...state,
+                hasFocus: true,
+                cursor: action.key
+            };
+        case 'UPDATE_SELECTION': {
+            const { key, selectionModel } = action;
+            return {
+                ...state,
+                selection: updateSelection(key, state.selection, selectionModel)
+            };
         }
+        default:
+            return state;
     }
-
-    handleBlur = () => {
-        this.setState(removeFocus);
-    };
-
-    getClickHandler = memoize(key => () => {
-        this.setState(updateSelection(key), this.syncSelection);
-    });
-
-    getFocusHandler = memoize(index => () => {
-        this.setState(updateCursor(index));
-    });
-}
+};
+
+const initialState = {
+    cursor: null,
+    hasFocus: false,
+    selection: new Set()
+};
+
+// TODO: move to a different file
+const useListState = ({ selectionModel, onSelectionChange }) => {
+    const [state, dispatch] = useReducer(reducer, initialState);
+    // when ever the selection changes, make the call
+    useEffect(() => {
+        onSelectionChange && onSelectionChange(state.selection);
+    }, [Array.from(state.selection).toString()]);
+    const removeFocus = useCallback(
+        () => dispatch({ type: 'REMOVE_FOCUS' }),
+        []
+    );
+    const updateSelection = key =>
+        useCallback(
+            () =>
+                dispatch({
+                    type: 'UPDATE_SELECTION',
+                    key,
+                    selectionModel
+                }),
+            []
+        );
+    const setFocus = key =>
+        useCallback(() => dispatch({ type: 'SET_FOCUS', key }), []);
+    return [state, { setFocus, removeFocus, updateSelection }];
+};
+
+const Items = props => {
+    const {
+        getItemKey,
+        items,
+        renderItem,
+        selectionModel,
+        onSelectionChange
+    } = props;
+    const [
+        { cursor, hasFocus, selection },
+        { setFocus: _setFocus, removeFocus, updateSelection: _updateSelection }
+    ] = useListState({ selectionModel, onSelectionChange });
+    return (
+        <Fragment>
+            {items.map((item, index) => {
```

Comment:
Avoid using multiline expressions in JSX; if you have to map some data to elements, try to define that as a `const` before `return`.

```jsx
const itemElements = items.map(item => <div />)

return (
  <Fragment>
    {itemElements}
  </Fragment>
)
```